Continual resizeobserver-driven flashing on Jupyter notebook page

smfr opened the pull request:
#### 912b33473fa3567ab5b1207a2b1230246ffa02ec
<pre>
Continual resizeobserver-driven flashing on Jupyter notebook page
<a href="https://bugs.webkit.org/show_bug.cgi?id=236346">https://bugs.webkit.org/show_bug.cgi?id=236346</a>
&lt;rdar://88399120&gt;

Reviewed by Alan Bujtas.

Based on a patch by Cathie Chen (cathiechen@igalia.com)

The page in question has a Resize Observer inside an `&lt;iframe&gt;` observing the `&lt;html&gt;` size, and
messages the enclosing document which attempts to size the iframe to its content. Line wrapping in
the iframe caused a vertical scrollbar to appear and disappear, triggering a continual flicker
between two states (when scrollbars are always visible).

The primary issue is that shrinking an iframe vertically would always trigger a vertical scrollbar,
even when the content could fit without one. This was a combination of two factors. First,
`ScrollView::setFrameRect()` updated scrollbars before forcing layout on the available content size
change, so `updateScrollbars()` saw a stale (too tall) document size and decided that a vertical
scrollbar was needed.

Second, `ScrollView::updateScrollbars()` would sometimes fail to trigger the layout that was
supposed to happen inside `updateContentsSize()`. This happened when there were no scrollbars. The
logic was attempting to avoid work for overlay scrollbars, but really needs to run if there&apos;s the
potential for a space-taking scrollbar to appear. For this, we add the virtual
`canShowNonOverlayScrollbars()` on ScrollableArea, and override on `FrameView` and
`RenderLayerScrollableArea`. Implementations determine if there can be any scrollbars, and if so
whether there is custom scrollbar style, or the theme specifies overlay scrollbars.

Finally, `ScrollView::setFrameRect()` only needs to call `updateScrollbars()` if it didn&apos;t call
`availableContentSizeChanged()`, since `updateScrollbars()` is called inside
`availableContentSizeChanged()`.

Other minor changes include use of SetForScope&lt;&gt;, and improvements to Layout logging.

* LayoutTests/fast/frames/iframe-resize-removes-scrollbar-expected.html: Added.
* LayoutTests/fast/frames/iframe-resize-removes-scrollbar.html: Added.
* LayoutTests/resize-observer/resize-observer-iframe-scrollbar-expected.txt: Added.
* LayoutTests/resize-observer/resize-observer-iframe-scrollbar.html: Added.
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::adjustScrollbarsForLayout):
(WebCore::FrameView::willDoLayout):
(WebCore::FrameView::canShowNonOverlayScrollbars const):
* Source/WebCore/page/FrameView.h:
* Source/WebCore/page/FrameViewLayoutContext.cpp:
(WebCore::FrameViewLayoutContext::unscheduleLayout):
(WebCore::FrameViewLayoutContext::layoutTimerFired):
* Source/WebCore/platform/ScrollTypes.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/ScrollView.cpp:
(WebCore::ScrollView::updateScrollbars):
(WebCore::ScrollView::setFrameRect):
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::canShowNonOverlayScrollbars const):
* Source/WebCore/platform/ScrollableArea.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::canShowNonOverlayScrollbars const):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:

Canonical link: <a href="https://commits.webkit.org/252440@main">https://commits.webkit.org/252440@main</a>
</pre>

cathiechen commented on LayoutTests/resize-observer/resize-observer-iframe-scrollbar.html:
```diff
@@ -0,0 +1,101 @@
+     <div id="console"></div>
+ </body>
+<script src="../resources/js-test-post.js"></script>
+ </html>
```
Needs an empty line here?

alanbaradlay commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -1488,6 +1492,19 @@ String FrameView::debugDescription() const
 
+bool FrameView::canShowNonOverlayScrollbars() const
+{
+    auto hasCustomScrollbarStyle = [&]() {
```
auto hasCustomScrollbarStyle = [&] ?

smfr committed:
Continual resizeobserver-driven flashing on Jupyter notebook page
https://bugs.webkit.org/show_bug.cgi?id=236346
<rdar://88399120>

Reviewed by Alan Bujtas.

Based on a patch by Cathie Chen (cathiechen@igalia.com)

The page in question has a Resize Observer inside an `<iframe>` observing the `<html>` size, and
messages the enclosing document which attempts to size the iframe to its content. Line wrapping in
the iframe caused a vertical scrollbar to appear and disappear, triggering a continual flicker
between two states (when scrollbars are always visible).

The primary issue is that shrinking an iframe vertically would always trigger a vertical scrollbar,
even when the content could fit without one. This was a combination of two factors. First,
`ScrollView::setFrameRect()` updated scrollbars before forcing layout on the available content size
change, so `updateScrollbars()` saw a stale (too tall) document size and decided that a vertical
scrollbar was needed.

Second, `ScrollView::updateScrollbars()` would sometimes fail to trigger the layout that was
supposed to happen inside `updateContentsSize()`. This happened when there were no scrollbars. The
logic was attempting to avoid work for overlay scrollbars, but really needs to run if there's the
potential for a space-taking scrollbar to appear. For this, we add the virtual
`canShowNonOverlayScrollbars()` on ScrollableArea, and override on `FrameView` and
`RenderLayerScrollableArea`. Implementations determine if there can be any scrollbars, and if so
whether there is custom scrollbar style, or the theme specifies overlay scrollbars.

Finally, `ScrollView::setFrameRect()` only needs to call `updateScrollbars()` if it didn't call
`availableContentSizeChanged()`, since `updateScrollbars()` is called inside
`availableContentSizeChanged()`.

Other minor changes include use of SetForScope<>, and improvements to Layout logging.

* LayoutTests/fast/frames/iframe-resize-removes-scrollbar-expected.html: Added.
* LayoutTests/fast/frames/iframe-resize-removes-scrollbar.html: Added.
* LayoutTests/resize-observer/resize-observer-iframe-scrollbar-expected.txt: Added.
* LayoutTests/resize-observer/resize-observer-iframe-scrollbar.html: Added.
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::adjustScrollbarsForLayout):
(WebCore::FrameView::willDoLayout):
(WebCore::FrameView::canShowNonOverlayScrollbars const):
* Source/WebCore/page/FrameView.h:
* Source/WebCore/page/FrameViewLayoutContext.cpp:
(WebCore::FrameViewLayoutContext::unscheduleLayout):
(WebCore::FrameViewLayoutContext::layoutTimerFired):
* Source/WebCore/platform/ScrollTypes.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/ScrollView.cpp:
(WebCore::ScrollView::updateScrollbars):
(WebCore::ScrollView::setFrameRect):
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::canShowNonOverlayScrollbars const):
* Source/WebCore/platform/ScrollableArea.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::canShowNonOverlayScrollbars const):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:

Canonical link: https://commits.webkit.org/252440@main
